controller/src/lib.rs

```diff
@@ -632,6 +747,215 @@ impl Controller {
     }
 
+    /// Set the power mode for a set of transceiver modules.
+    ///
+    /// This method may be used regardless of whether a module uses hardware
+    /// control or software override for controlling the power.
+    pub async fn set_power_mode(&self, modules: ModuleId, mode: PowerMode) -> Result<(), Error> {
+        // How we proceed largely depends on two things: whether we're turning
+        // the power OFF entirely, and whether a module has set software
+        // override of the `LPMode` pin.
+        match mode {
+            PowerMode::Off => {
+                self.assert_lpmode(modules).await?;
+                self.assert_reset(modules).await?;
+                self.disable_power(modules).await
+            }
+            other => {
+                // Validate the power state transition.
+                //
+                // For now, we enforce that modules may not go directly to high
+                // power, they have to go through low-lower first.
+                //
+                // We can always set modules to low power, though, since that's
+                // valid from both off and high-power, and a no-op if it's
+                // already set.
+                let current_power_state = self.power_mode(modules).await?;
+                if matches!(mode, PowerMode::High) {
+                    if current_power_state
+                        .iter()
+                        .any(|(mode, _override)| mode == &PowerMode::Off)
+                    {
+                        return Err(Error::InvalidPowerStateTransition);
+                    }
+                }
+
+                // We need the status bits to determine if we also need to
+                // twiddle `ResetL`.
+                let status = self.status(modules).await?;
+
+                // Check whether power is enabled / good, and / or reset
+                // asserted for any of the requested modules. We need to manage
+                // those pin states to control the power. Note that this is true
+                // regardless of whether the module has software-override of
+                // power control set. That's because we want the pins and the
+                // memory map to reflect the same state, so that toggling the
+                // software override doesn't change the power state of the
+                // module, only which hardware signals it responds to.
+                let need_power_enabled = Self::filter_modules_with(modules, &status, |st| {
+                    !st.contains(Status::POWER_GOOD | Status::ENABLED)
+                })?;
+                let need_reset_deasserted =
+                    Self::filter_modules_with(modules, &status, |st| st.contains(Status::RESET))?;
+                if need_power_enabled.selected_transceiver_count() > 0 {
+                    self.enable_power(need_power_enabled).await?;
+                }
+                if need_reset_deasserted.selected_transceiver_count() > 0 {
```

Comment:
Is it possible for some transceivers to already be **not in reset**, but have power disabled?

I believe that would be a violation of our intended sequencing, but should we detect / warn about it / handle it specially?

Reply:
I think this is something we should check for since we there are timing requirements around reset, particularly the first reset after power on ("device initialization").

Reply:
I would not be surprised if interactions with the SDE left a transceiver in that sort of state, though it'd be a bit weird. That would mean the SDE put a transceiver in reset, and then some of our software powered-off the module, using the direct interface like `xcvradm disable-power`.

I think it should be fairly easy to at least warn about this case, with a few bit operations on the `PortMask`s. I'll try it out, and either add it now if it's easy or make an issue. Thanks for bringing it up!

Reply:
@mkeeter @Aaron-Hartwig Fixed this up in 4cfceaf. I've made this so that, if we are enabling power on any modules which _do not_ already have `ResetL` asserted, we:

- Emit a warning
- Assert `ResetL`
- Wait for SFF-8679 `t_reset_init` (10us)
- For all modules being powered (even those which previously did not have `ResetL` asserted, i.e., the odd ones), then assert `ResetL`
- Wait for SFF-8679 `t_reset` (2s)

Reply:
That behavior now looks like this:

```bash
ben@niles ~ $ ./xcvradm -i axf2 -t0 status
FPGA Port Status
   0    0 PRESENT | INTERRUPT
ben@niles ~ $ ./xcvradm -i axf2 -t0 set-power low
Jan 31 20:59:46.840 WARN Found modules with power disabled, but reset deasserted. It will be asserted before enabling power., suspicious_modules: ModuleId { fpga_id: 0, ports: PortMask(1) }, need_reset_deasserted: ModuleId { fpga_id: 0, ports: PortMask(0) }, need_power_enabled: ModuleId { fpga_id: 0, ports: PortMask(1) }
ben@niles ~ $ ./xcvradm -i axf2 -t0 status
FPGA Port Status
   0    0 PRESENT | ENABLED | LOW_POWER_MODE | INTERRUPT | POWER_GOOD
ben@niles ~ $ ./xcvradm -i axf2 -t0 power
FPGA  Port  Power  Software-override
   0     0  Low    No
ben@niles ~ $
```